src/rime/gear/script_translator.cc

```diff
@@ -301,34 +361,48 @@ bool ScriptTranslation::Evaluate(Dictionary* dict, UserDictionary* user_dict) {
 }
 
 bool ScriptTranslation::Next() {
-  if (exhausted())
-    return false;
-  if (sentence_) {
-    sentence_.reset();
-    return !CheckEmpty();
-  }
-  int user_phrase_code_length = 0;
-  if (user_phrase_ && user_phrase_iter_ != user_phrase_->rend()) {
-    user_phrase_code_length = user_phrase_iter_->first;
-  }
-  int phrase_code_length = 0;
-  if (phrase_ && phrase_iter_ != phrase_->rend()) {
-    phrase_code_length = phrase_iter_->first;
-  }
-  if (user_phrase_code_length > 0 &&
-      user_phrase_code_length >= phrase_code_length) {
-    DictEntryList& entries(user_phrase_iter_->second);
-    if (++user_phrase_index_ >= entries.size()) {
-      ++user_phrase_iter_;
-      user_phrase_index_ = 0;
+  bool is_correction;
+  do {
+    is_correction = false;
+    if (exhausted())
+      return false;
+    if (sentence_) {
+      sentence_.reset();
+      return !CheckEmpty();
     }
-  }
-  else if (phrase_code_length > 0) {
-    DictEntryIterator& iter(phrase_iter_->second);
-    if (!iter.Next()) {
-      ++phrase_iter_;
+    int user_phrase_code_length = 0;
+    if (user_phrase_ && user_phrase_iter_ != user_phrase_->rend()) {
+      user_phrase_code_length = user_phrase_iter_->first;
     }
-  }
+    int phrase_code_length = 0;
+    if (phrase_ && phrase_iter_ != phrase_->rend()) {
+      phrase_code_length = phrase_iter_->first;
+    }
+    if (user_phrase_code_length > 0 &&
+        user_phrase_code_length >= phrase_code_length) {
+      DictEntryList& entries(user_phrase_iter_->second);
+      if (++user_phrase_index_ >= entries.size()) {
+        ++user_phrase_iter_;
+        user_phrase_index_ = 0;
+      }
+    }
+    else if (phrase_code_length > 0) {
+      DictEntryIterator& iter(phrase_iter_->second);
+      if (!iter.Next()) {
+        ++phrase_iter_;
+      }
+    }
+    if (enable_correction_) {
+      PrepareCandidate();
+      if (!candidate_) {
+        break;
+      }
+      is_correction = syllabifier_->IsCandidateCorrection(*candidate_);
+    }
+  } while ( // limit the number of correction candidates
+      enable_correction_ && is_correction &&
```

Comment:
nit: break into two lines
to make it more obvious how many conditions are chained, use different lines for every item if one line cannot contain all items in the same level.